src/graph/network.cc

```diff
@@ -606,5 +635,99 @@ DGL_REGISTER_GLOBAL("network._CAPI_DeleteKVMsg")
     delete msg;
   });
 
+DGL_REGISTER_GLOBAL("network._CAPI_FastPull")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+    std::string name = args[0];
+    int local_machine_id = args[1];
+    int machine_count = args[2];
+    int group_count = args[3];
+    int client_id = args[4];
+    NDArray ID = args[5];
+    NDArray pb = args[6];
+    NDArray g2l = args[7];
+    NDArray local_data = args[8];
+    CommunicatorHandle chandle_sender = args[9];
+    CommunicatorHandle chandle_receiver = args[10];
+    network::Sender* sender = static_cast<network::Sender*>(chandle_sender);
+    network::Receiver* receiver = static_cast<network::SocketReceiver*>(chandle_receiver);
+    int64_t ID_size = ID.GetSize() / sizeof(int64_t);
+    int64_t* ID_data = static_cast<int64_t*>(ID->data);
+    int64_t* pb_data = static_cast<int64_t*>(pb->data);
+    int64_t* g2l_data = static_cast<int64_t*>(g2l->data);
+    char* local_data_char = static_cast<char*>(local_data->data);
+    std::vector<int64_t> local_ids;
+    std::vector<int64_t> local_ids_orginal;
+    std::vector<int64_t> local_data_shape;
+    std::vector<std::vector<int64_t> > remote_ids(machine_count);
+    std::vector<std::vector<int64_t> > remote_ids_original(machine_count);
+    int row_size = 1;
+    for (int i = 0; i < local_data->ndim; ++i) {
+      local_data_shape.push_back(local_data->shape[i]);
+      if (i != 0) {
+        row_size *= local_data->shape[i];
+      }
+    }
+    row_size *= sizeof(float);
+    // Get local id and remote id
+    for (int64_t i = 0; i < ID_size; ++i) {
+      int64_t id = ID_data[i];
+      int64_t part_id = pb_data[id];
+      if (part_id == local_machine_id) {
+        int64_t local_id = g2l_data[id];
+        local_ids.push_back(local_id);
+        local_ids_orginal.push_back(i);
+      } else {
+        remote_ids[part_id].push_back(id);
+        remote_ids_original[part_id].push_back(i);
+      }
+    }
+    int msg_count = 0;
+    for (int i = 0; i < remote_ids.size(); ++i) {
+      if (remote_ids[i].size() != 0) {
+        KVStoreMsg kv_msg;
+        kv_msg.msg_type = MessageType::kPullMsg;
+        kv_msg.rank = client_id;
+        kv_msg.name = name;
+        kv_msg.id = CreateNDArrayFromRaw({static_cast<int64_t>(remote_ids[i].size())},
+                                         DLDataType{kDLInt, 64, 1},
+                                         DLContext{kDLCPU, 0},
+                                         remote_ids[i].data(),
+                                         !AUTO_FREE);
+        send_kv_message(sender, &kv_msg, i, !AUTO_FREE);
```

Comment:
when is the memory from CreateNDArrayFromRaw free'd? after it exit from the code block?